billing/lib/api.ts

```diff
@@ -0,0 +1,290 @@
+
+export type SpaceSnapshotStore =
+  & StorePutter<SpaceSnapshot>
+  & StoreGetter<SpaceSnapshotKey, SpaceSnapshot>
+
+/**
+ * Captures information about a customer of the service that may need to be
+ * billed for storage usage.
+ */
+export interface Customer {
+  /** CID of the UCAN invocation that set it to the current value. */
+  cause: Link,
+  /** DID of the user account e.g. `did:mailto:agent`. */
+  customer: DID<'mailto'>,
+  /**
+   * Opaque identifier representing an account in the payment system
+   * e.g. Stripe customer ID (stripe:cus_9s6XKzkNRiz8i3)
+   */
+  account: string,
+  /** Unique identifier of the product a.k.a tier. */
+  product: string,
+  /** Time the record was added to the database. */
+  insertedAt: Date,
+  /** Time the record was updated in the database. */
+  updatedAt: Date
+}
+
+export interface CustomerListOptions extends Pageable {}
+
+export type CustomerStore = StoreLister<{}, Customer>
+
+/**
+ * Captures storage usage by a given customer for a given space in the given
+ * time period.
+ */
+export interface Usage {
+  /** Customer DID (did:mailto:...). */
+  customer: DID<'mailto'>
+  /**
+   * Opaque identifier representing an account in the payment system.
+   * 
+   * e.g. Stripe customer ID (stripe:cus_9s6XKzkNRiz8i3)
+   */
+  account: string
+  /** Unique identifier of the product a.k.a tier. */
+  product: string
+  /** Storage provider for the space. */
+  provider: DID<'web'>
+  /** Space DID (did:key:...). */
+  space: DID
+  /** Usage in byte/ms */
+  usage: bigint
+  /** Time the usage period spans from (inclusive). */
+  from: Date
+  /** Time the usage period spans to (exclusive). */
+  to: Date
+  /** Time the record was added to the database. */
+  insertedAt: Date
+}
+
+export interface UsageKey { customer: DID<'mailto'>, from: Date }
+
+export type UsageStore = StorePutter<Usage>
+
+// Billing queues /////////////////////////////////////////////////////////////
+
+/**
+ * Captures details about a customer that should be billed for a given period
+ * of usage.
+ */
+export interface CustomerBillingInstruction {
+  /** Customer DID (did:mailto:...). */
+  customer: DID<'mailto'>
+  /**
+   * Opaque identifier representing an account in the payment system.
+   * 
+   * e.g. Stripe customer ID (stripe:cus_9s6XKzkNRiz8i3)
+   */
+  account: string
+  /** Unique identifier of the product a.k.a tier. */
+  product: string
+  /** Time the billing period spans from (inlusive). */
+  from: Date
+  /** Time the billing period spans to (exclusive). */
+  to: Date
+}
+
+export type CustomerBillingQueue = QueueAdder<CustomerBillingInstruction>
+
+/**
+ * Captures details about a space that should be billed for a given customer in
+ * the given period of usage.
+ */
+export interface SpaceBillingInstruction extends CustomerBillingInstruction {
+  /** Space DID (did:key:...). */
+  space: DID
+  /** Storage provider for the space. */
+  provider: DID<'web'>
```

Comment:
Could be a good idea to define `type Provider = DID<'web'>` and use that, same for `Consumer`  and `Customer` that way it reads more clearly and we can update definitions in one place when necessary.